In [66]:
# **************************************************************************** #
#                                                                              #
#                                                         :::      ::::::::    #
#                                                       :+:      :+:    :+:    #
#                                                     +:+ +:+         +:+      #
#    By: luicasad <luicasad@student.42.fr>          +#+  +:+       +#+         #
#                                                 +#+#+#+#+#+   +#+            #
#    Created: 2023/10/22 11:17:16 by luicasad          #+#    #+#              #
#    Updated: 2023/10/22 11:17:24 by luicasad         ###   ########.fr        #
#                                                                              #
# **************************************************************************** #

<span style='color:red'> Augmenta l'amplada del quadern jupyter<br> </span><span style='color:green'> Aumento el ancho del cuaderno Jupyter<br></span><span style='color:blue'> Increase the width of jupyter notebook <br></span>

In [67]:
from IPython.display import display, HTML
display(HTML("<style>.container { width:100%}</style>"))

In [68]:
import os
import pandas as pd
import geopandas as gpd
import numpy as np
home_dir = os.getcwd()
data_folder =  home_dir +'/../../../data/accid/'

In [69]:
def	print_head_tail(df):
    print(df.iloc[[0,1, 2, df.index[-3],df.index[-2],df.index[-1]]])

def concat_files(data_folder:str, years: range, common_name:str )-> pd.DataFrame:

    #For each year, 
    #   create a dataframe reading a csv file.
    #   Append it to a list
    #   Create a dataframe with columns count

    files = []
    counts =[]
    for year in years:
        file = data_folder + str(year) + common_name
        df = pd.read_csv(file)
        files.append(df)
        counts.append(pd.DataFrame(data=df.count(), columns=[year]))
        print(year)
        print(df.shape)
        print(list(df))
        #print(df.head(3))
        print("-" * 80)
    print(pd.concat(counts, axis=1))
    
    return pd.concat(files, ignore_index=True)

def count_NAN(df):
    """
    Count NAN values in each column of a dataset.
    Print 8 counts in per line
    """
    col_counter = 1
    for col in df.columns:
        count = df[col].isna().sum()
        if ((col_counter % 8) == 0):
            sep = "|\n"
        else:
            sep = "| "
        print("|{} = {}".format(col,count), end=sep)
        col_counter = col_counter + 1
def minus_one(df):
    """
    Count NAN values in each column of a dataset.
    Print 8 counts in per line
    """
    col_counter = 1
    for col in df.columns[:-1]:
        count = df[df[col] == -1][col].count()
        if ((col_counter % 8) == 0):
            sep = "|\n"
        else:
            sep = "| "
        print("|{} = {}".format(col,count), end=sep)
        col_counter = col_counter + 1

def concat_accid_files(data_folder:str, \
                       years: range, \
                       common_name:str, \
                       headers:list)-> pd.DataFrame:

    #For each year, 
    #   create a dataframe reading a csv file.
    #   Append it to a list
    #   Create a dataframe with columns count
    field_types = {'Codi_districte':int, \
                   'Codi_barri':int, \
                   'Codi_carrer':int, \
                   'NK_Any':int, \
                   'Mes_any':int, \
                   'Dia_mes':int, \
                   'Coordenada_UTM_X_ED50':np.float64, \
                   'Coordenada_UTM_Y_ED50':np.float64, \
                   'Longitud':np.float64, \
                   'Latitud':np.float64}
    converters = {"Longitud": lambda x: np.nan if x == '' else x, \
                  "Longitud": lambda x: np.nan if x == '' else x}
    files = []
    counts =[]
    for year in years:
        file = data_folder + str(year) + common_name
        df = pd.read_csv(file, dtype=field_types)
        df.columns = headers
        files.append(df)
        counts.append(pd.DataFrame(data=df.count(), columns=[year]))
        #print(year)
        #print(df.shape)
        #print(df.head(3))
        #print("-" * 80)
    #print(pd.concat(counts, axis=1))
    
    return pd.concat(files, ignore_index=True)
    
def save_preprocessed(df,data_folder:str,file_name:str):
	print("Saving preprocessed data.....")

	# File path
	file_path = data_folder + file_name

	# Save data
	df.to_csv(file_path, index= False)

<span style='color:red'> Llegir i concatenar tots els fitxers d'accidents com Pandas GEO DataFrames<br> </span><span style='color:green'> Lectura y concatenación de los archivos de accidentes como Pandas GEO DataFrames<br></span><span style='color:blue'> Read and concatenate all accident files as Pandas GEO DataFrames <br></span>

In [70]:
pd.set_option("display.width", 120)
accid = concat_files(data_folder, range(2016, 2023),"_accidents_causa_conductor_gu_bcn_.csv")
#pd.concat(counts, axis=1)
print_head_tail(accid)

2016
(10510, 20)
['Numero_expedient', 'Codi_districte', 'Nom_districte', 'Codi_barri', 'Nom_barri', 'Codi_carrer', 'Nom_carrer', 'Num_postal ', 'Descripcio_dia_setmana', 'NK_Any', 'Mes_any', 'Nom_mes', 'Dia_mes', 'Hora_dia', 'Descripcio_torn', 'Descripcio_causa_conductor', 'Coordenada_UTM_X_ED50', 'Coordenada_UTM_Y_ED50', 'Longitud', 'Latitud']
--------------------------------------------------------------------------------
2017
(11091, 20)
['Numero_expedient', 'Codi_districte', 'Nom_districte', 'Codi_barri', 'Nom_barri', 'Codi_carrer', 'Nom_carrer', 'Num_postal ', 'Descripcio_dia_setmana', 'NK_Any', 'Mes_any', 'Nom_mes', 'Dia_mes', 'Hora_dia', 'Descripcio_torn', 'Descripcio_causa_conductor', 'Coordenada_UTM_X_ED50', 'Coordenada_UTM_Y_ED50', 'Longitud', 'Latitud']
--------------------------------------------------------------------------------
2018
(10835, 20)
['Numero_expedient', 'Codi_districte', 'Nom_districte', 'Codi_barri', 'Nom_barri', 'Codi_carrer', 'Nom_carrer', 'Num_postal ', 

<span style='color:red'> Podem apreciar l'intercanvi d'informació entre el torn (Descripcio_torn ) i la causa (Descripcio_causa_conductor) i l'existència de dos noms de camps per a la causa (Descripcio_causa_conductor, Descripcio_causa_mediata): SOlucionem el problema aplicant una capçalera<br> </span><span style='color:green'> Podemos apreciar el intercambio de informacion entre el turno y la causa y la existencia de dos nombres de campos para la causa: SOlucionamoes el problema aplicando una cabecera comun a todos los archivos.<br></span><span style='color:blue'> We can see the exchange of information between the shift (Descripcio_torn) and the cause (Descripcio_causa_conductor) and the existence of two field names for the cause (Descripcio_causa_conductor, Descripcio_causa_mediata): We solve the problem by applying a common header to all files. <br></span>

In [71]:
headers= ['Numero_expedient', 'Codi_districte', 'Nom_districte', 'Codi_barri', 'Nom_barri', \
          'Codi_carrer', 'Nom_carrer', 'Num_postal ', 'Descripcio_dia_setmana', 'NK_Any', \
          'Mes_any', 'Nom_mes', 'Dia_mes', 'Hora_dia', 'Causa', 'Torn', \
          'Coordenada_UTM_X_ED50', 'Coordenada_UTM_Y_ED50', 'Longitud', 'Latitud']
accid = concat_accid_files(data_folder, range(2016, 2023),"_accidents_causa_conductor_gu_bcn_.csv", headers)


In [72]:
#print(accid.dtypes)

In [73]:
print("Original Dataset has {} rows with {} columns.".format(accid.shape[0],accid.shape[1]))

Original Dataset has 69355 rows with 20 columns.


### <span style='color:red'> Estudi dels valors Nan<br> </span><span style='color:green'> Estudio de los valores nulos<br></span><span style='color:blue'> Study Nan values <br></span>

In [74]:
count_NAN(accid)

|Numero_expedient = 0| |Codi_districte = 0| |Nom_districte = 0| |Codi_barri = 0| |Nom_barri = 0| |Codi_carrer = 0| |Nom_carrer = 1| |Num_postal  = 16|
|Descripcio_dia_setmana = 0| |NK_Any = 0| |Mes_any = 0| |Nom_mes = 0| |Dia_mes = 0| |Hora_dia = 0| |Causa = 0| |Torn = 0|
|Coordenada_UTM_X_ED50 = 0| |Coordenada_UTM_Y_ED50 = 0| |Longitud = 30| |Latitud = 30| 

In [75]:
minus_one(accid)

|Numero_expedient = 0| |Codi_districte = 501| |Nom_districte = 0| |Codi_barri = 501| |Nom_barri = 0| |Codi_carrer = 646| |Nom_carrer = 0| |Num_postal  = 0|
|Descripcio_dia_setmana = 0| |NK_Any = 0| |Mes_any = 0| |Nom_mes = 0| |Dia_mes = 0| |Hora_dia = 0| |Causa = 0| |Torn = 0|
|Coordenada_UTM_X_ED50 = 30| |Coordenada_UTM_Y_ED50 = 30| |Longitud = 0| 

<span style='color:red'> Els 30 registres sense coordenades UTM tenen als camps Longitud i Latitud valors NAN i el barri i el carrer desconegut. Els esborrem.<br> </span><span style='color:green'> Los 30 registros sin coordenas UTM tiene en los campos Longitud y Latitud valores NAN y el barrio y la calle desconocida. Los borramos. <br></span><span style='color:blue'> The 30 records without UTM coordinates have NAN values and the unknown neighborhood and street in the Longitude and Latitude fields. We delete them. <br></span>

In [76]:
drop_mask = (accid["Coordenada_UTM_X_ED50"] == -1) | (accid["Coordenada_UTM_Y_ED50"] == -1)
accid[drop_mask][["Nom_barri", "Nom_carrer","Coordenada_UTM_X_ED50", "Coordenada_UTM_Y_ED50", "Longitud",	"Latitud"]].head(30)


Nom_barri  Nom_carrer  Coordenada_UTM_X_ED50  Coordenada_UTM_Y_ED50  Longitud  Latitud
52129  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
52130  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
52131  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
52132  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
52133  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
52134  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
52135  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
53298  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
53299  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
54395  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
54396  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
54397  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
54398  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
55521  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
55522  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
55523  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
55524  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
55525  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
55526  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
56660  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
57819  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
57820  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
57821  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
57822  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
58944  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
58945  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
58946  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
58947  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
58948  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN
58949  Desconegut  Desconegut                   -1.0                   -1.0       NaN      NaN

In [77]:
accid.drop(accid[drop_mask].index, inplace = True)
print("Dataset Without neighborhoods with NaNs has {} rows with {} columns.".format(accid.shape[0],accid.shape[1]))

Dataset Without neighborhoods with NaNs has 69325 rows with 20 columns.


<span style='color:red'> Vegem si tenim dades geogràfiques als barris desconeguts.<br> </span><span style='color:green'> Veamos si tenemos datos geograficos en los barrios desconocidos. <br></span><span style='color:blue'> Let's see if we have geographic data in the unknown neighborhoods. <br></span>

In [78]:
drop_mask = (accid["Codi_barri"] == -1) 
accid[drop_mask][["Nom_barri", "Nom_carrer","Coordenada_UTM_X_ED50", "Coordenada_UTM_Y_ED50", "Longitud",	"Latitud"]].head(5).sort_values(by="Longitud")


Nom_barri                                         Nom_carrer  Coordenada_UTM_X_ED50  Coordenada_UTM_Y_ED50  \
1281  Desconegut                                         Desconegut              426830.03             4576250.78   
1     Desconegut  ALMIRALL CERVERA / Sevilla                    ...              432316.21             4581276.81   
2     Desconegut  NÚMERO 3 / E                                  ...              427265.97             4576644.81   
3     Desconegut  RDA DALT (BESÒS)                              ...              424991.14             4581760.18   
0     Desconegut  NÚMERO 6 / A                                  ...              427519.99             4575229.36   

      Longitud    Latitud  
1281  2.124450  41.332574  
1     2.154737  41.376367  
2     2.157427  41.381284  
3     2.157427  41.381284  
0     2.165891  41.375166

In [79]:
accid[drop_mask][["Nom_barri", "Nom_carrer","Coordenada_UTM_X_ED50", "Coordenada_UTM_Y_ED50", "Longitud",	"Latitud"]].tail(5).sort_values(by="Longitud")

Nom_barri                                         Nom_carrer  Coordenada_UTM_X_ED50  Coordenada_UTM_Y_ED50  \
68190  Desconegut  Augusta                                       ...              426355.48             4584211.14   
68193  Desconegut  Joan XXIII                                    ...              426451.50             4581657.08   
68774  Desconegut  Arístides Maillol / Cardenal Reig             ...              426465.56             4581445.06   
68191  Desconegut  Corts / Comte de Güell                        ...              426676.50             4581334.04   
68192  Desconegut  Guipúscoa                                     ...              433835.76             4586344.61   

       Longitud    Latitud  
68190  2.117812  41.404227  
68193  2.119271  41.381232  
68774  2.119465  41.379324  
68191  2.122001  41.378343  
68192  2.207062  41.424093

In [80]:
#Create a copy of records with unkown neighborhoods

unkown_barris = accid[drop_mask]
unkown_barris.shape
#remove records with with unkown neighborhoods from original data set.
accid.drop(accid[drop_mask].index, inplace = True)

<span style='color:red'> Intentarem inferir els noms i els codis dels barris a partir de les coordenades UTM i d'un shapefile amb els barris de la ciutat.<br> </span><span style='color:green'> Intentaremos inferir los nombres y códigos de los barrios a partir de las coordenadas UTM y de un shapefile con los barrios de la ciudad.<br></span><span style='color:blue'> We will try to infer the names and codes of the neighborhoods from the UTM coordinates and a shapefile with the city's neighborhoods. <br></span>

In [81]:
import geopandas as gpd
bcn_barris_shapes = gpd.read_file('/home/luis/Documentos/python/jump2digital/data/shapefiles/bcn_barris', encoding='utf-8')
print("The coordinate reference system for this shapefile is {}".format(bcn_barris_shapes.crs))

The coordinate reference system for this shapefile is EPSG:25831


In [82]:
#let's simplify shapefile to hold only fields interesting for this job.
bcn_barris_shapes = bcn_barris_shapes[[ 'DISTRICTE', 'BARRI', 'NOM', 'geometry']]

In [83]:
#with utm coordinates create a POINT() as the geometry field
unkown_barris.loc[:,"geometry"] = gpd.GeoSeries.from_xy(x=unkown_barris.Coordenada_UTM_X_ED50, y=unkown_barris.Coordenada_UTM_Y_ED50, crs="EPSG:25831")
#Transform DataFrame into a GeoDataFrame
geo_unkown_barris = gpd.GeoDataFrame(unkown_barris, geometry='geometry')
geo_unkown_barris.shape

/tmp/ipykernel_10685/2452988879.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  unkown_barris.loc[:,"geometry"] = gpd.GeoSeries.from_xy(x=unkown_barris.Coordenada_UTM_X_ED50, y=unkown_barris.Coordenada_UTM_Y_ED50, crs="EPSG:25831")


(471, 21)

In [84]:
#Look for shapes that contain the accident geo point. It mades a left join, regularly resulti less records cause some of them do not fit in any shape.
found_barris=gpd.sjoin(geo_unkown_barris, bcn_barris_shapes, predicate = 'within')
found_barris.shape


(434, 25)

<span style='color:red'> Hem tingut èxit amb 434 registres d'accidents..<br> </span><span style='color:green'> Hemos tenido éxito con 434 registros de accidentes.<br></span><span style='color:blue'> We have been successful with 434 accident records. <br></span>

In [85]:
#Copy information from Shapefile's fields to correspondign accid-dataset's field
found_barris["Codi_districte"]=found_barris["DISTRICTE"]
found_barris["Codi_barri"]=found_barris["BARRI"]
found_barris["Nom_barri"]=found_barris["NOM"]
#remove joined columns
found_barris.drop("geometry", axis=1, inplace=True)
found_barris.drop("index_right", axis=1, inplace=True)
found_barris.drop("DISTRICTE", axis=1, inplace=True)
found_barris.drop("BARRI", axis=1, inplace=True)
found_barris.drop("NOM", axis=1, inplace=True)

found_barris.shape


(434, 20)

In [86]:
#append recovered data into the original dataset.
accid = pd.concat([accid, found_barris], ignore_index=True)


In [87]:
minus_one(accid)


|Numero_expedient = 0| |Codi_districte = 0| |Nom_districte = 0| |Codi_barri = 0| |Nom_barri = 0| |Codi_carrer = 579| |Nom_carrer = 0| |Num_postal  = 0|
|Descripcio_dia_setmana = 0| |NK_Any = 0| |Mes_any = 0| |Nom_mes = 0| |Dia_mes = 0| |Hora_dia = 0| |Causa = 0| |Torn = 0|
|Coordenada_UTM_X_ED50 = 0| |Coordenada_UTM_Y_ED50 = 0| |Longitud = 0| 

In [88]:
print("Dataset Without neighborhoods with NaNs has {} rows with {} columns.".format(accid.shape[0],accid.shape[1]))

Dataset Without neighborhoods with NaNs has 69288 rows with 20 columns.


### <span style='color:red'> Agregació trimestral dels accidents<br> </span><span style='color:green'> Agregación trimestral de los accidentes <br></span><span style='color:blue'> Quarterly aggregation of accidents <br></span>

In [89]:
#Create a column for quarters
accid=accid.assign(
    Trimestre=accid[["Mes_any"]].apply(
        lambda row: ((row -1)//3)+1),axis=1)
#create a key column to merge the results into the price table
accid=accid.assign(
    Clau=accid[["Codi_barri", "NK_Any", "Trimestre"]].apply(
        lambda row: ''.join([str(each) for each in row]),axis=1))

In [90]:
count_NAN(accid)

|Numero_expedient = 0| |Codi_districte = 0| |Nom_districte = 0| |Codi_barri = 0| |Nom_barri = 0| |Codi_carrer = 0| |Nom_carrer = 1| |Num_postal  = 15|
|Descripcio_dia_setmana = 0| |NK_Any = 0| |Mes_any = 0| |Nom_mes = 0| |Dia_mes = 0| |Hora_dia = 0| |Causa = 0| |Torn = 0|
|Coordenada_UTM_X_ED50 = 0| |Coordenada_UTM_Y_ED50 = 0| |Longitud = 0| |Latitud = 0| |Trimestre = 0| |axis = 0| |Clau = 0| 

In [91]:
grouped_accid = accid.groupby("Clau",group_keys=False)
counted_quarter = grouped_accid.size()

In [92]:
grouped_accid = accid.groupby(["Clau","Torn"],group_keys=False)
counted_quarted_torn = grouped_accid.size().unstack()

In [93]:
counted_quarted_torn.fillna(0, inplace=True)

In [94]:
mat = [x[0] for x in counted_quarted_torn.values]
nit = [x[1] for x in counted_quarted_torn.values]
tar = [x[2] for x in counted_quarted_torn.values]
mat_perc = [x[0]/sum(x) for x in counted_quarted_torn.values]
nit_perc = [x[1]/sum(x) for x in counted_quarted_torn.values]
tar_perc = [x[2]/sum(x) for x in counted_quarted_torn.values]

In [95]:
accidents_by_quarter_and_torn=pd.DataFrame({"Clau":counted_quarter.index, \
                                             "Accidents":counted_quarter.values, \
                                             "Mat":mat, \
                                             "Tar":tar, \
                                             "Nit":nit, \
                                             "Mat_P":mat_perc, \
                                             "Tar_p":tar_perc, \
                                             "Nit_p":nit_perc,})

In [96]:
accidents_by_quarter_and_torn


Clau  Accidents   Mat   Tar   Nit     Mat_P     Tar_p     Nit_p
0     0320163          1   1.0   0.0   0.0  1.000000  0.000000  0.000000
1     0320193          1   0.0   1.0   0.0  0.000000  1.000000  0.000000
2     0420173          1   0.0   1.0   0.0  0.000000  1.000000  0.000000
3     0420182          1   0.0   1.0   0.0  0.000000  1.000000  0.000000
4     0420183          1   1.0   0.0   0.0  1.000000  0.000000  0.000000
...       ...        ...   ...   ...   ...       ...       ...       ...
2041   920214         95  37.0  44.0  14.0  0.389474  0.463158  0.147368
2042   920221         72  27.0  35.0  10.0  0.375000  0.486111  0.138889
2043   920222         75  27.0  42.0   6.0  0.360000  0.560000  0.080000
2044   920223         87  37.0  35.0  15.0  0.425287  0.402299  0.172414
2045   920224         88  34.0  42.0  12.0  0.386364  0.477273  0.136364

[2046 rows x 8 columns]

In [97]:
save_preprocessed(accidents_by_quarter_and_torn, data_folder, "accidents_trimestrals_16_22.csv")
accidents_by_quarter_and_torn.shape

Saving preprocessed data.....


(2046, 8)